### PR TITLE
contrail logs

### PR DESCRIPTION
**Pull request by anshprat**
pointing to my repo for testing

### CODE REVIEW

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

File: .fixtures.yml [hunk 1]
```diff
@@ -32,7 +32,8 @@ fixtures:
       repo: https://github.com/saz/puppet-ssh
       ref: v2.3.6
     contrail:
-      repo: https://github.com/jiocloud/jiocloud-contrail
+      repo: https://github.com/anshprat/jiocloud-contrail
+      ref: logs_1
```
Comment (anshprat) [inline .fixtures.yml]: These are tmp changes for testing, to be reverted before merge after jiocloud-contrail merge

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
@hkumar @soren @bodepd This is ready for review

File: files/contrail-consolidate-logs.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
```
Comment (hkumarmk) [inline files/contrail-consolidate-logs.sh]: What is the motive of this script to move the errors and other stuffs to daily log?

File: hiera/data/common.yaml [hunk 3]
```diff
@@ -643,6 +643,26 @@ contrail::enable_analytics: false
 
 rjil::contrail::server::enable_analytics: false
 
+log_level: SYS_INFO
```
Comment (hkumarmk) [inline hiera/data/common.yaml]: All these params are default params in contrail module, so no need to pass them through hiera, let them use class defaults.

File: files/contrail-consolidate-logs.sh (hunk 2 above)
Comment (anshprat) [inline files/contrail-consolidate-logs.sh]: These logs are presently size bound to the tune of few MBs. Hence presently we only have a couple of hours worth of data. So we first aggregate the data to a daily file and let logrotate handle the daily file. Our codebase at this stage does not allow logging config modifications for most of contrail modules. Upstream is now slighty better (patched april end) but not all modules are covered yet.

File: manifests/contrail/server.pp [hunk 4]
```diff
@@ -20,18 +20,35 @@
...
-                      'contrail-collector',
-                      'query-engine',
-                      'api',
-                      'discovery',
-                      'schema',
-                      'svc-monitor',
-                      'contrail-control',
-                      'webserver',
-                      'jobserver',
+  $contrail_logs = ['contrail-api-daily',
```
Comment (hkumarmk) [inline manifests/contrail/server.pp]: I am bit confused about these "daily" logs, why do we need them like this?

File: manifests/contrail/logrotate/consolidate.pp [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class rjil::contrail::logrotate::consolidate {
```
Comment (hkumarmk) [inline manifests/contrail/logrotate/consolidate.pp]: why do we need to consolidate the logs? can't it be done through log configuration?

File: manifests/jiocloud/logrotate.pp [hunk 6]
```diff
@@ -1,14 +1,17 @@
 define rjil::jiocloud::logrotate(
   $logdir        = "/var/log",
-  $logfile       = "notset",
+  $logfile       = 'undef',
```
Comment (hkumarmk) [inline manifests/jiocloud/logrotate.pp]: undef without quotes

File: manifests/jiocloud/logrotate.pp [hunk 7]
```diff
@@ -1,14 +1,17 @@
...
   $rotate        = 60,
   $compress      = true,
   $delaycompress = true,
   $ifempty       = false,
+  $copytruncate  = 'undef',
+  $dateext       = true,
+  $ensure        = 'present',
 ) {
-  if ($logfile == "notset"){
+  if ($logfile == 'undef'){
```
Comment (hkumarmk) [inline manifests/jiocloud/logrotate.pp]: you just need if !$logfile

File: manifests/jiocloud/logrotate.pp [hunk 8]
```diff
@@ -1,14 +1,17 @@
...
       if ($logdir =~ /\/$/) {
```
Comment (hkumarmk) [inline manifests/jiocloud/logrotate.pp]: This is not required, there is no harm to just add ${logdir}/${name}.log" irrespective of logdir is end with '/' or not

File: manifests/contrail/server.pp [hunk 9]
```diff
@@ -20,18 +20,35 @@
...
+  ]
+  
+  rjil::jiocloud::logrotate { $contrail_logs_copytruncate:
+    logdir       => '/var/log/contrail',
+    copytruncate => true,
+  }
+  
+  include rjil::contrail::logrotate::consolidate
+
+  $contrail_logrotate_delete = ['contrail-config',
```
Comment (hkumarmk) [inline manifests/contrail/server.pp]: What is this? why you need to remove certain logrotate entries?

File: manifests/contrail/logrotate/consolidate.pp (hunk 5 above)
Comment (anshprat) [inline manifests/contrail/logrotate/consolidate.pp]: Our codebase at this stage does not allow logging config modifications for most of contrail modules.

**Comment by hkumarmk** [comment]
@anshprat It would be better you add some inline documentation to brief what and why those codes.

File: files/contrail-consolidate-logs.sh (hunk 2 above)
Comment (bodepd) [inline files/contrail-consolidate-logs.sh]: can we not just make the desired adjustments to the logrotate config? This seems really complicated.

Comment (anshprat) [inline files/contrail-consolidate-logs.sh]: @bodepd unfortunately logrotate will only work on top of a given logfile. The issue is with the given logfile presently which is only of few MBs here. So if we just keep rotating them, for retaining 60 days of logs, each day will generate upwards of 24 logs :! Or use two logrotate configs - one to create something like the daily log version, and another to rotate the daily log. Hence creating this script to create a standard file which can be just rotated by logrotate.

File: manifests/jiocloud/logrotate.pp (hunk 8 above)
Comment (anshprat) [inline manifests/jiocloud/logrotate.pp]: yeah, not required from practical purpose but added this to keep the config file sames. Else depending upon how the param is provided, will end up with configs having double slashes.

File: manifests/contrail/server.pp (hunk 9 above)
Comment (anshprat) [inline manifests/contrail/server.pp]: These default config files provided by package have multiple log lines in each config file. Hence some of them override the configs we are creating here. So deleting these default config files.

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

File: files/contrail-consolidate-logs.sh (hunk 2 above)
Comment (bodepd) [inline files/contrail-consolidate-logs.sh]: I'll have a close look at this tomorrow. My instinct is that the fact that you need to model things as a procedure using bash means that you aren't thinking about the problem right (or at least in the Puppet way). I feel like there must be a way to model things as the desired state of the configuration. I can spend some time with a build tomorrow to get in depth enough to try to better understand your motivations here and if there might be a better way.

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by bodepd** [comment]
@anshprat if you get a build working that you feel good about, please leave it for me, and I'll jump on and take a look when I wake up.

**Comment by anshprat** [comment]
if only I had a cent for every failed build... well, atleast I could ve a free lunch :/

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
@bodepd 1321 is all green.

**Comment by bodepd** [comment]
It might be better if I try to catch you online, I don't quite understand what is going on here :(

Let me try to state what I think the issue is that leads to all of this complexity:

contrail uses log4j as the backend for it's logging, and is also using log4j to manage log rotation. This rotation is based on log file size, and for some services is set so low, that a majority of the log information can't be retained in a way that support our requirements.

To resolve this, you are running a script continuously to append the log contents into a daily log file to each service, then specifying log rotation rules on that file.

To offer an alternative that might be a little easier to implement, can we just disable the logrotation that is happening through the contrail config (maybe by setting it to -1 or something) and then rely entirely on logrotation rules? I'm sure that I'm still missing some details about why that is not possible. I'll try to catch up to you during your Mon or Tues morning.

**Comment by bodepd** [comment]
one more thing, I went through all of the config files and resulting log files by hand, and lots of the log files indicated contrail errors. I cross checked the log files with prod and did not see the same log entries there. I can also cross check against a recent build to see if the errors exist there as well. If these changes are causing contrail to be in a failed state, then our validation scripts are not robust enough in this case to detect it :(

failures in the dns logs:

````
sh: Send: FAILED: 1432926490533535 IFMapStateMachine [SYS_DEBUG]: IFMapSmEventMessage: Processing ifsm::EvConnectionCleaned in state SsrcStart controller/src/ifmap/client/ifmap_state_machine.cc 944
2015-05-29 Fri 19:08:10:533.608 UTC  ct1-testjenkins-puppet-rjil-gate-1321 [Thread 140132601620224, Pid 26806]: Sandesh: Send: FAILED: 1432926490533596 IFMapStateMachine [SYS_DEBUG]: IFMapSmTransitionMessage: SsrcStart ===> ConnectTimerWait controller/src/ifmap/client/ifmap_state_machine.cc 954
2015-05-29 Fri 19:08:10:533.653 UTC  ct1-testjenkins-puppet-rjil-gate-1321 [Thread 140132601620224, Pid 26806]: Sandesh: Send: FAILED: 1432926490533640 IFMapStateMachine [SYS_DEBUG]: IFMapSmStartTimerMessage: 1 second connect timer started. controller/src/ifmap/client/ifmap_state_machine.cc 707
````

/var/log/contrail/contrail-discovery.log
````
05/29/2015 06:54:58 PM [ct1-testjenkins-puppet-rjil-gate-1321:DiscoveryService:Config:0]: No Connection: __default__ [SYS_INFO]: discServiceLog: subscribe: service type=Collector, client=VRouterAgent:cp2-testjenkins-puppet-rjil-gate-1321:VRouterAgent, ttl=32, asked=2 pubs=0/0, subs=0
````

/var/log/contrail/contrail-schema.log

````
05/29/2015 11:01:27 AM [ct1-testjenkins-puppet-rjil-gate-1321:Schema:Config:0]: Sandesh Client: Event[EvCollectorUnknown] => State[Connect] -> State[Disconnect]
05/29/2015 11:02:22 AM [ct1-testjenkins-puppet-rjil-gate-1321:Schema:Config:0]: Processing event[EvSandeshUVESend] in state[Disconnect]
````
control log also has lots of failures:
````
desh: Send: FAILED: 1432927174918091 IFMapStateMachine [SYS_WARN]: IFMapPeerConnError: wrong version number SsrcConnect SsrcSslHandshake ifsm::EvConnectSuccess controller/src/ifmap/client/ifmap_state_machine.cc 916
2015-05-29 Fri 19:19:34:918.154 UTC  ct1-testjenkins-puppet-rjil-gate-1321 [Thread 140144828499904, Pid 26953]: Sandesh: Send: FAILED: 1432927174918146 IFMapStateMachine [SYS_DEBUG]: IFMapSmEventMessage: Enqueuing ifsm::EvHandshakeFailed in state SsrcSslHandshake controller/src/ifmap/client/ifmap_state_machine.cc 937
2015-05-29 Fri 19:19:34:918.222 UTC  ct1-testjenkins-puppet-rjil-gate-1321 [Thread 140144660563712, Pid 26953]: Sandesh: Send: FAILED: 1432927174918210 IFMapStateMachine [SYS_DEBUG]: IFMapSmEventMessage: Processing ifsm::EvHandshakeFailed in state SsrcSslHandshake controller/src/ifmap/client/ifmap_state_machine.cc 944
2015-05-29 Fri 19:19:34:918.258 UTC  ct1-testjenkins-puppet-rjil-gate-1321 [Thread 140144660563712, Pid 26953]: Sandesh: Send: FAILED: 1432927174918250 IFMapStateMachine [SYS_DEBUG]: IFMapSmCancelTimerMessage: Cancelling Response timer. controller/src/ifmap/client/ifmap_state_machine.cc 743
2015-05-29 Fri 19:19:34:918.290 UTC  ct1-testjenkins-puppet-rjil-gate-1321 [Thread 140144660563712, Pid 26953]: Sandesh: Send: FAILED: 1432927174918282 IFMapStateMachine [SYS_DEBUG]: IFMapSmTransitionMessage: SsrcSslHandshake ===> SsrcStart controller/src/ifmap/client/ifmap_state_machine.cc 954
2015-05-29 Fri 19:19:34:918.489 UTC  ct1-testjenkins-puppet-rjil-gate-1321 [Thread 140144828499904, Pid 26953]: Sandesh: Send: FAILED: 1432927174918477 IFMapPeer [SYS_DEBUG]: IFMapServerConnection: Retrying connection to Ifmap-server.  controller/src/ifmap/client/ifmap_channel.cc 204
````

ifmap server:

````
2015-05-29 19:23:50,925 [pool-6-thread-97] ERROR  - ChannelThread: Setting channel 10.0.0.98:47203:20507 into state 'BROKEN'
2015-05-29 19:23:53,921 [pool-6-thread-97] ERROR  - ChannelThread: Receiving request failed
2015-05-29 19:23:53,922 [pool-6-thread-97] ERROR  - ChannelThread: SSLHandshakeException: Client doesn't know about our certificate (?)
2015-05-29 19:23:53,922 [pool-6-thread-97] ERROR  - ChannelThread: Setting channel 10.0.0.98:47218:20508 into state 'BROKEN'
2015-05-29 19:23:53,933 [pool-6-thread-97] ERROR  - ChannelThread: Receiving request failed
2015-05-29 19:23:53,933 [pool-6-thread-97] ERROR  - ChannelThread: SSLHandshakeException: Client doesn't know about our certificate (?)
2015-05-29 19:23:53,933 [pool-6-thread-97] ERROR  - ChannelThread: Setting channel 10.0.0.98:47219:20509 into state 'BROKEN'
````

**Comment by anshprat** [comment]
@bodepd lets sync up on this today?

**Comment by anshprat** [comment]
please test this

File: files/contrail-consolidate-logs.sh [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+		'schema.err'
+		'api-0-zk'
+		'schema-zk'
+		)
+
+log_dir='/var/log/contrail/'
+for log in ${contrail_logs[@]} 
+do 
+file="${log_dir}${log}"
+file_log="${file}.log.1"
```
Comment (bodepd) [inline files/contrail-consolidate-logs.sh]: I just logged in and had a look, this is the piece of data that I was missing before, you are checking the rotated out files. I feel like there is some risk that we might lost log contents here, I'm going to leave a build alive for a day and see what the results look like.

All in all, it doesn't seem as scary in terms of race-conditions as I thought. I'm going to sit down and think about possible race conditions, but for the most part, I think it's okay.

**Comment by bodepd** [comment]
I did some testing and found two issues:

1. the files are saved as root and not as the contrail user (I'm not sure if this matters, but it makes the inconsistent with the rest of the logs)
2. in any case where it adds a number higher than 1, those files are not appended (ie: .2, .3). I know given how often the script runs, this should not happen, but I think we still need logic to protect against this case.

**Comment by anshprat** [comment]
1) - Agreed about consistency. Though I can fix up the ownership as part of this PR, I think fixing the  ownership as part of fixing all (contrail + other) logs to be non-sudo readable would be consistent and easier.
2) - I ve knowingly ignored .2+ since they will come into picture only when .1 is lying around. When this script is first pushed there will be some of these older logs. Adding them to the consolidation script will add complexity and will add a lot of cases to be handled. 

For example - lets say, someone decompresses one of the older pre-script log.5.gz files in the same location. This newly uncompressed log.5 should not be added  to the daily.log. Again, I can add check for time etc, but this adds complexity. 

Another thing to be take care of would be to ensure that the logs are appended to daily in right chronological order.

There is one usecase where it could be useful - this custom script somehow did not run correctly/was paused manually for long enough for .2 to be created.

Logrotate script would generally handle only the specified log file. For the purpose of this script, the specified log file is log.1.

So we have to take a call whether to handle all such situations. Its doable, but I am not sure its worth the various issues that can crop up due to this or worth the complexity it may add. Thoughts?

**Comment by bodepd** [comment]
let's merge it.

**Pull request merged**